// clarity-abi/dist/cjs/types/descrimination.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isClarityAbiList = exports.isClarityAbiTuple = exports.isClarityAbiOptional = exports.isClarityAbiResponse = exports.isClarityAbiStringUtf8 = exports.isClarityAbiStringAscii = exports.isClarityAbiBuffer = exports.isClarityAbiPrimitive = exports.narrow = void 0;
function narrow(value) {
    return value;
}
exports.narrow = narrow;
const isClarityAbiPrimitive = (val) => typeof val === 'string';
exports.isClarityAbiPrimitive = isClarityAbiPrimitive;
const isClarityAbiBuffer = (val) => val.buffer !== undefined;
exports.isClarityAbiBuffer = isClarityAbiBuffer;
const isClarityAbiStringAscii = (val) => val['string-ascii'] !== undefined;
exports.isClarityAbiStringAscii = isClarityAbiStringAscii;
const isClarityAbiStringUtf8 = (val) => val['string-utf8'] !== undefined;
exports.isClarityAbiStringUtf8 = isClarityAbiStringUtf8;
const isClarityAbiResponse = (val) => val.response !== undefined;
exports.isClarityAbiResponse = isClarityAbiResponse;
const isClarityAbiOptional = (val) => val.optional !== undefined;
exports.isClarityAbiOptional = isClarityAbiOptional;
const isClarityAbiTuple = (val) => val.tuple !== undefined;
exports.isClarityAbiTuple = isClarityAbiTuple;
const isClarityAbiList = (val) => val.list !== undefined;
exports.isClarityAbiList = isClarityAbiList;
