// ts-clarity/dist/esm/lib/network/request.js
import fetch from 'cross-fetch';
import { kDefaultExtendedApiRequestTimeout, kDefaultStacksMainnetEndpoint, kDefaultStacksReadRetries, kDefaultStacksTimeout, } from '../common/constants.js';
import { exponentialBackoff, } from '../common/fetch.js';
export function mergeDefaultExtendedApiRequestOptions(options) {
    return {
        stacksEndpoint: options?.stacksEndpoint ?? kDefaultStacksMainnetEndpoint,
        timeout: options?.timeout ?? kDefaultExtendedApiRequestTimeout,
        fetch: options?.fetch ?? fetch,
        retries: options?.retries ?? kDefaultStacksReadRetries,
        retryDelay: options?.retryDelay ?? exponentialBackoff(),
    };
}
export function mergeDefaultNodeApiRequestOptions(options) {
    return {
        stacksEndpoint: options?.stacksEndpoint ?? kDefaultStacksMainnetEndpoint,
        timeout: options?.timeout ?? kDefaultStacksTimeout,
        fetch: options?.fetch ?? fetch,
        retries: options?.retries ?? kDefaultStacksReadRetries,
        retryDelay: options?.retryDelay ?? exponentialBackoff(),
    };
}
