// ts-clarity/dist/cjs/lib/network/info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNodeInfo = getNodeInfo;
exports.getExtendedStatus = getExtendedStatus;
const fetch_js_1 = require("../common/fetch.js");
const request_js_1 = require("./request.js");
async function getNodeInfo(_options) {
    const options = (0, request_js_1.mergeDefaultNodeApiRequestOptions)(_options);
    const url = `${options.stacksEndpoint}/v2/info`;
    const rs = await (0, fetch_js_1.richFetch)(url, {
        timeout: options.timeout,
        fetch: options.fetch,
        retries: options.retries,
        retryDelay: options.retryDelay,
        retryOn: fetch_js_1.retryOnError,
    });
    return await rs.json();
}
async function getExtendedStatus(_options) {
    const options = (0, request_js_1.mergeDefaultExtendedApiRequestOptions)(_options);
    const url = `${options.stacksEndpoint}/extended/v1/status`;
    const rs = await (0, fetch_js_1.richFetch)(url, {
        timeout: options.timeout,
        fetch: options.fetch,
        retries: options.retries,
        retryDelay: options.retryDelay,
        retryOn: fetch_js_1.retryOnError,
    });
    const status = await rs.json();
    if (status.status !== 'ready')
        return null;
    return status;
}
