// ts-clarity/dist/cjs/lib/network/request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mergeDefaultExtendedApiRequestOptions = mergeDefaultExtendedApiRequestOptions;
exports.mergeDefaultNodeApiRequestOptions = mergeDefaultNodeApiRequestOptions;
const cross_fetch_1 = require("cross-fetch");
const constants_js_1 = require("../common/constants.js");
const fetch_js_1 = require("../common/fetch.js");
function mergeDefaultExtendedApiRequestOptions(options) {
    return {
        stacksEndpoint: options?.stacksEndpoint ?? constants_js_1.kDefaultStacksMainnetEndpoint,
        timeout: options?.timeout ?? constants_js_1.kDefaultExtendedApiRequestTimeout,
        fetch: options?.fetch ?? cross_fetch_1.default,
        retries: options?.retries ?? constants_js_1.kDefaultStacksReadRetries,
        retryDelay: options?.retryDelay ?? (0, fetch_js_1.exponentialBackoff)(),
    };
}
function mergeDefaultNodeApiRequestOptions(options) {
    return {
        stacksEndpoint: options?.stacksEndpoint ?? constants_js_1.kDefaultStacksMainnetEndpoint,
        timeout: options?.timeout ?? constants_js_1.kDefaultStacksTimeout,
        fetch: options?.fetch ?? cross_fetch_1.default,
        retries: options?.retries ?? constants_js_1.kDefaultStacksReadRetries,
        retryDelay: options?.retryDelay ?? (0, fetch_js_1.exponentialBackoff)(),
    };
}
