// ts-clarity/dist/cjs/lib/contract/read-variable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.readVariable = readVariable;
const transactions_1 = require("@stacks/transactions");
const assert_js_1 = require("../common/assert.js");
const constants_js_1 = require("../common/constants.js");
const fetch_js_1 = require("../common/fetch.js");
const buffer_js_1 = require("../utils/buffer.js");
const abi_codec_js_1 = require("./abi-codec.js");
const constants_js_2 = require("./constants.js");
async function readVariable(params) {
    const { stacksEndpoint, proof, indexBlockHash, contract, variableName, abi } = params;
    const varDef = abi.find((def) => def.name === variableName);
    (0, assert_js_1.assert)(varDef != null, `failed to find variable definition for ${variableName}`);
    const urlParams = new URLSearchParams();
    urlParams.set('proof', proof === true ? '1' : '0');
    if (typeof indexBlockHash === 'string') {
        urlParams.set('tip', indexBlockHash.substring(2).toLowerCase());
    }
    else if (indexBlockHash instanceof Uint8Array) {
        urlParams.set('tip', (0, buffer_js_1.fromUint8Array)(indexBlockHash).toString('hex'));
    }
    const [deployer, contractName] = contract.split('.', 2);
    const vn = String(variableName);
    const url = `${stacksEndpoint ?? constants_js_2.kDefaultStacksEndpoint}/v2/data_var/${deployer}/${contractName}/${vn}?${urlParams.toString()}`;
    const funcResponse = await (0, fetch_js_1.richFetch)(url, {
        method: 'GET',
        fetch: params.fetch,
        timeout: params.timeout ?? constants_js_1.kDefaultStacksTimeout,
        retries: params.retries ?? constants_js_1.kDefaultStacksReadRetries,
        retryDelay: params.retryDelay ?? (0, fetch_js_1.exponentialBackoff)(),
        retryOn: fetch_js_1.retryOnError,
    });
    if (funcResponse.status !== 200) {
        throw new Error(`Read contract variable failed with http code ${funcResponse.status}: ${funcResponse.statusText}`);
    }
    const mapResult = await funcResponse.json();
    const result = (0, transactions_1.deserializeCV)(mapResult.data);
    return (0, abi_codec_js_1.decodeAbi)(varDef.type, result);
}
