// clarity-abi/dist/cjs/known-abis/ownable-trait.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = {
    functions: [
        {
            name: 'set-contract-owner',
            access: 'public',
            args: [
                {
                    name: 'owner',
                    type: 'principal',
                },
            ],
            outputs: {
                type: {
                    response: {
                        ok: 'bool',
                        error: 'uint128',
                    },
                },
            },
        },
        {
            name: 'get-contract-owner',
            access: 'read_only',
            args: [],
            outputs: {
                type: {
                    response: {
                        ok: 'principal',
                        error: 'none',
                    },
                },
            },
        },
    ],
    variables: [],
    maps: [],
    fungible_tokens: [],
    non_fungible_tokens: [],
    epoch: 'Epoch2_05',
    clarity_version: 'Clarity1',
};
